### ISSUE TITLE
Feature request: {1..5} as syntax for toList(1..5)

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1978]
Currently we have:
```m2
i1 : 1..5

o1 = (1, 2, 3, 4, 5)

o1 : Sequence

i2 : 5:1

o2 = (1, 1, 1, 1, 1)

o2 : Sequence
```
This is a proposal to support:
```m2
i3 : {1..5}

o3 = {1, 2, 3, 4, 5} -- instead of {(1, 2, 3, 4, 5)}

o3 : List

i4 : {5:1}

o4 = {1, 1, 1, 1, 1}  -- instead of {(1, 1, 1, 1, 1)}

o4 : List
```
This is a slight breaking change, since to get the previous output one would have to do:
```m2
i5 : {(1..5)}

o5 = {(1, 2, 3, 4, 5)}

o5 : List

i6 : {(5:1)}

o6 = {(1, 1, 1, 1, 1)}

o6 : List
```
But this is an easier and much more intuitive syntax, and avoids the hassle of having to write `toList(1..5)` or `toList(5:1)`.

As an extension, we can probably make the following work as well:
```m2
i7 : [1..5]

o7 = [1, 2, 3, 4, 5]

o7 : Array

i8 : <|1..5|>

o8 = <|1, 2, 3, 4, 5|>

o8 : AngleBarList
```
Like I said, I think this is much more intuitive, so I think it's worth the trouble of a slight breaking change. We've had similar breaking changes before, for instance changing the commenting syntax.

**Comment by DanGrayson** [comment]
That is more intuitive syntax, good idea.

**Comment by mahrud** [comment]
Any hints as to where might this be implemented? For instance, how does this happen:
```m2
i1 : 1,2,3

o1 = (1, 2, 3)

o1 : Sequence

i2 : {1,2,3}

o2 = {1, 2, 3}

o2 : List
```
What prevents this from giving `{(1,2,3)}`?

**Comment by DanGrayson** [comment]
It's probably in `convertr.d`, here:
```
     is p:Parentheses do (
	  if p.left.word == leftparen then convert(p.contents)
	  else if p.left.word == leftbrace 
	  then Code(listCode(makeCodeSequence(p.contents,CommaW),treePosition(e)))
	  else 
	  if p.left.word == leftbracket 
	  then Code(arrayCode(makeCodeSequence(p.contents,CommaW),treePosition(e)))
	  else 
	  if p.left.word == leftAngleBar
	  then Code(angleBarListCode(makeCodeSequence(p.contents,CommaW),treePosition(e)))
	  else 
	  dummyCode			  -- should not happen
	  )
```

**Comment by mahrud** [comment]
What does `convertr` mean?

**Comment by DanGrayson** [comment]
That file **converts** parse trees to executable pseudocode.

**Comment by mahrud** [comment]
What is the "r" for?

**Comment by DanGrayson** [comment]
I'm guessing it's short for "converter".

**Comment by pzinn** [comment]
As much as I find writing `toList(1..5)` annoying, I'm a bit afraid of what this change would do.
I mean, logically, `{1..5}` *should* produce a list containing the sequence `(1,2,3,4,5)`. This is what the syntax is saying. If you start redefining it, then with the same logic should we declare that `{anySequence}` is effectively the same as `toList(anySequence)`?

**Comment by mahrud** [comment]
> I mean, logically, `{1..5}` _should_ produce a list containing the sequence `(1,2,3,4,5)`. This is what the syntax is saying.

The [documentation](http://www2.macaulay2.com/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_...html) literally says:
> The most confusing thing about this operator, in all its guises, is that it is not a syntactic construction

I don't understand the rest of your claim. I'm only suggestion that the `..` and `:` become part of the syntax and respect user's choice for what type they should return. With your reasoning, do you expect entering `{1,2,3,4,5}` to return `{(1,2,3,4,5)}`?

**Comment by mahrud** [comment]
The [documentation](http://www2.macaulay2.com/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/___Z__Z_sp_co_sp__Thing.html) for `ZZ : Thing` also has this warning:
> Warning: such sequences do not get automatically spliced into lists containing them.

I think simplifying the syntax and removing this warning would make the language more intuitive.

**Comment by mahrud** [comment]
I was thinking I would implement this by adding a new `FlatList` type that get's immediately absorbed by the container it is in, so for instance `{FlatList{1,2,3}, FlatList{4,5,6}}` would turn into `{1,2,3,4,5,6}`, but I'm not sure how to prevent something like `select(1..6, even)` from turning into `select(1,2,3,4,5,6,even)`!

One solution would be to have flattening occur on assignment (e.g. when it is being assigned to `o2`), but I would need to look more into the interpreter to see if that can work.

I thought about also combining this with #1904 and returning an iterator which behaves as above. @d-torrance do you have any thoughts on this? I think it would be nice to have `{10:0}..{10:10}` return a nested interator instead of a list with 10^10 pairs!

**Comment by DanGrayson** [comment]
I implemented that idea of having something like FlatList in Mathematica, whose instances get spliced into lists automatically, but it turned out to be a bad idea.  It's too kludgy.

Making {1..5} expand to {1,2,3,4,5} at parsing time is fine, but at run time, things should follow a model of computation that is simple and principled.

**Comment by mahrud** [comment]
> Making {1..5} expand to {1,2,3,4,5} at parsing time is fine, but at run time, things should follow a model of computation that is simple and principled.

I think I figured it out: inside types that inherit from List, we get `{1..3,5..7} -> {1,2,3,5,6,7}` but inside types that inherit from Sequence we the same combination yields `((1,2,3), (5,6,7))`, which can then be spliced. This would guarantee that `select(1..3, even)` turns into `select((1,2,3), even)` and not `select(1,2,3,even)`.

I really like this. It satisfies that "Mathematician's intuition" for sets, so you would have:
```m2
i1 : set{1..3,5..7}

o1 = set {1, 2, 3, 5, 6, 7}

o1 : Set
```
While not breaking how functions work in Macaulay2.

**Comment by DanGrayson** [comment]
I'm not sure I like `{1..3}` being different from `x=1..3; {x}`.  Probably the new syntax should be with 3 periods, to distinguish it from the binary operator.

**Comment by mahrud** [comment]
What do you not like about that? Doesn't it make sense given this behavior?
```m2
i1 : 1,2,3,4

o1 = (1, 2, 3, 4)

o1 : Sequence

i2 : {oo}

o2 = {(1, 2, 3, 4)}

o2 : List

i3 : {1,2,3,4}

o3 = {1, 2, 3, 4}

o3 : List
```
Typing `x=1..3` is implicitly typing `x=(1,2,3)`, while typing `x={1..3}` is implicitly typing `x={1,2,3}`. I definitely don't want to convert any sequence `(...)` wrapped by braces into a list. That's what Paul was worried about, and I definitely agree that it would cause issues.

I think adding `...` would unnecessarily complicate the language more, and we would have to teach people that now there's this new operator, etc.

**Comment by DanGrayson** [comment]
The syntax is totally clear at the moment: `..` is a binary operator, just like `+` or `-`, and at run time an expression such as `x..y` is evaluated and returns a value, which gets used like any other value.  The comma, on the other hand, is part of the syntax, and that's why it behaves the way it does -- it's handled at parsing time, not at run time.  What you want is syntax.

**Comment by mahrud** [comment]
Yes, I'm not confused about that. I'm suggesting changing `{1..4,6..10}` to be handled at parsing time as a part of the syntax.

**Comment by mahrud** [comment]
Oh, to be clear, when I said "inside types that inherit from List", I didn't mean this would be handled at the top level. I just meant in `convertr.d` we differentiate between `leftparen` and `leftbrace,leftbracket,leftAngleBar`.

**Comment by mahrud** [comment]
It would be good to keep `..` as a symbol so people could implement other things like `{X..Y}` where `X` and `Y` are objects of some type. Maybe the `FlatList` idea isn't so bad. Here is a suggestion:
- During parsing `{1..3, 4, 2:5}` is treated by replacing `1..3` with a `FlatList` containing a caller for `(symbol.., ZZ, ZZ)` and `2:5` with a `FlatList` containing a caller for `(symbol:, ZZ, Thing)`
- At runtime, the flat lists are evaluated and `{FlatList(1..3),4,FlatList(2:5)}` is turned into `{1,2,3,4,5,5}`.

The difference being that for the sequence `(1..3,4,2:5)`, Sequence is used instead of FlatList.

**Comment by DanGrayson** [comment]
That seems sort of kludgey to me.

**Comment by mahrud** [comment]
Again, the point is to simplify the language and get the notation closer to mathematics so it's easier for mathematicians to use it.

**Comment by DanGrayson** [comment]
Yes, I realize that's the aim, but it seems to make the language more complicated, instead.

**Comment by mikestillman** [comment]
By the way, @DanGrayson and I years ago tried hard to set up the notation 1..n, r:1, to be part of lists, but could not find a good way to do it.

**Comment by mahrud** [comment]
I'll reiterate this suggestion, since I think it is really the most ideal:

> I think I figured it out: inside types that inherit from List, we get `{1..3,5..7} -> {1,2,3,5,6,7}` but inside types that inherit from Sequence the same combination yields `((1,2,3), (5,6,7))`, which can then be spliced if desired. This would guarantee that `select(1..3, even)` turns into `select((1,2,3), even)` and not `select(1,2,3,even)`, while `select({1..3,5..7}, even)` still does the sensible thing.
> 
> I really like this. It satisfies that "Mathematician's intuition" for sets, so you would have:
> 
> ```m2
> i1 : set{1..3,5..7}
> 
> o1 = set {1, 2, 3, 5, 6, 7}
> 
> o1 : Set
> ```
> 
> While not breaking how functions work in Macaulay2.

**Comment by mahrud** [comment]
> And currently this can be done with this alternative, which works fine:
> 
> ```m2
> i8 : splice {1..3, 4, 2:5}
> 
> o8 = {1, 2, 3, 4, 5, 5}
> 
> o8 : List
> ```

@DanGrayson your solution doesn't scale:
```m2
i3 : splice {(0,0)..(1,1), (2,2), 2:(3,3)}

o3 = {(0, 0), (0, 1), (1, 0), (1, 1), 2, 2, (3, 3), (3, 3)}

o3 : List
```

**Comment by mahrud** [comment]
Related: #2020
